api/server/v1/error.go

```diff
@@ -17,98 +17,292 @@ package api
 type TigrisError struct {
 	// The status code, which should be an enum value of [google.rpc.Code][google.rpc.Code]. We don't need to marshal
 	// this code for HTTP clients.
-	Code codes.Code `json:"code,omitempty"`
+	Code Code `json:"code,omitempty"`
 	// A developer-facing error message.
 	Message string `json:"message,omitempty"`
-	// A list of messages that carry the error details. This is mainly to send our internal codes and messages.
-	Details []*ErrorDetails `json:"details,omitempty"`
+
+	Details []proto.Message `json:"details,omitempty"`
 }
 
 // Error to return the underlying error message
 func (e *TigrisError) Error() string {
 	return e.Message
 }
 
-// GRPCStatus converts the TigrisError and return status.Status. This is used to return grpc status to the grpc clients
-func (e *TigrisError) GRPCStatus() *status.Status {
-	s := &spb.Status{
-		Code:    int32(e.Code),
-		Message: e.Message,
-	}
+// WithDetails a helper method for adding details to the TigrisError
+func (e *TigrisError) WithDetails(details ...proto.Message) *TigrisError {
+	e.Details = append(e.Details, details...)
+	return e
+}
 
-	if len(e.Details) == 0 {
-		return status.FromProto(s)
+// WithRetry attached retry information to the error
+func (e *TigrisError) WithRetry(d time.Duration) *TigrisError {
+	if d != 0 {
+		e.Details = append(e.Details, &errdetails.RetryInfo{RetryDelay: durationpb.New(d)})
 	}
+	return e
+}
+
+// RetryDelay retrieves retry delay if it's attached to the error
+func (e *TigrisError) RetryDelay() time.Duration {
+	var dur time.Duration
 
-	var details []*any.Any
 	for _, d := range e.Details {
-		var a = &any.Any{}
-		err := anypb.MarshalFrom(a, d, proto.MarshalOptions{})
-		if err == nil {
-			details = append(details, a)
+		switch t := d.(type) {
+		case *errdetails.RetryInfo:
+			dur = t.RetryDelay.AsDuration()
 		}
 	}
 
-	s.Details = details
-	return status.FromProto(s)
+	return dur
 }
 
-// WithDetails a helper method for adding details to the TigrisError
-func (e *TigrisError) WithDetails(details ...*ErrorDetails) *TigrisError {
-	e.Details = details
-	return e
+// ToGRPCCode converts Tigris error code to GRPC code
+// Extended codes converted to 'Unknown' GRPC code
+func ToGRPCCode(code Code) codes.Code {
+	if code <= Code_UNAUTHENTICATED {
+		return codes.Code(code)
+	}
+
+	// Cannot be represented by GRPC codes
+	return codes.Unknown
+}
+
+// ToTigrisCode converts GRPC code to Tigris code
+func ToTigrisCode(code codes.Code) Code {
+	return Code(code)
+}
+
+func CodeFromString(c string) Code {
+	code, ok := Code_value[c]
+	if !ok {
+		return Code_UNKNOWN
+	}
+
+	return Code(code)
+}
+
+func StringFromCode(c Code) string {
+	r, ok := Code_name[int32(c)]
+	if !ok {
+		r = Code_name[int32(Code_UNKNOWN)]
+	}
+
+	return r
+}
+
+// ToHTTPCode converts Tigris' code to HTTP status code
+// Used to customize HTTP codes returned by GRPC-gateway
+func ToHTTPCode(code Code) int {
+	switch code {
+	case Code_OK:
+		return 200
```

Comment:
Can we use the constants defined in the HTTP package?

Reply:
Many of the codes below doesn't have HTTP representation. Partially replacing them wouldn't look nice, I think.